File: sorting/merge_sort/mergesort.c
#include <stdio.h>
#define begin 0
#define end 5

void merge(int *vector, int begin, int middle, int end){
	int size1, size2, i, j;

	size1 = middle - begin + 1;
	size2 = end - middle;

	int e[size1], d[size2];

	for(i=0;i<size1;i++){
		e[i] = vector[begin + i];
	}

	for(j=0;j<size2;j++){
		d[j] = vector[middle+1+j];
	}
	int k = begin;
	i = 0;
	j = 0;

	while(i<size1 && j<size2){
		if(e[i]<=d[j]){
			vector[k] = e[i];
			i++;
		}
		else{
			vector[k] = d[j];
			j++;
		}
		k++;
	}
	
	while(i<size1){
		vector[k] = e[i];
		k++;
		i++;
	}
	
	while(j<size2){
		vector[k] = d[j];
		j++;
		k++;
	}
}

void merge_sort(int *vector, int begin, int end){
	
	int middle = begin + (end - begin)/2;
	
	if(begin < end){
		merge_sort(vector,begin, middle);
		merge_sort(vector,middle+1, end);
		merge(vector,begin, middle, end);
	}
}

File: sorting/merge_sort/mergesort.php
<?php    
$unsorted = array(43,21,2,1,9,24,2,99,23,8,7,114,92,5);

    function merge_sort($array)

    {

    $length = count($array);

    if($length <= 1)

    {

    return $array;

    }

    $left = $right = $result = array();

    $middle = round($length / 2,0);

    for($i=0;$i<$middle;$i++)

    {

    $left[] = $array[$i];

    }

    for($j=$middle;$j<$length;$j++)

    {

    $right[] = $array[$j];

    }

    $left = merge_sort($left);

    $right = merge_sort($right);

    $result = merge($left, $right);

    return $result;

    }

    function merge($left, $right)

    {

    $result = array();

    while(count($left) > 0 || count($right) > 0)

    {

    if(count($left) > 0 && count($right) > 0)

    {

    if($left[0] <= $right[0]){

    // get first item from array and shift items over

    $result[] = array_shift($left);

    }

    else{

    $result[] = array_shift($right);

    }

    }

    elseif(count($left) > 0){

    $result[] = array_shift($left);

    }

    else{

    $result[] = array_shift($right);

    }

    }

    return $result;

    }

    $sorted = merge_sort($unsorted);

    print_r($sorted);
?>
